homeassistant/components/apple_tv/__init__.py

```diff
@@ -70,205 +70,257 @@ def ensure_list(value: Union[T, Sequence[T]]) -> Sequence[T]:
+        self._start_connect_loop()
+
+    def connection_closed(self):
+        """Device connection was (intentionally) closed."""
+        self.atv = None
+        self._update_state(disconnected=True)
+        self._start_connect_loop()
+
+    async def connect(self):
+        """Connect to device."""
+        self._is_on = True
+        self._start_connect_loop()
+
+    async def disconnect(self):
+        """Disconnect from device."""
+        _LOGGER.debug("Disconnecting from device")
+        self._is_on = False
+        try:
+            if self.atv:
+                self.atv.push_updater.listener = None
+                self.atv.push_updater.stop()
+                await self.atv.close()
+                self.atv = None
+            if self._task:
+                self._task.cancel()
+                self._task = None
+        finally:
+            self._update_state(disconnected=False)
+
+    def _start_connect_loop(self):
+        if not self._task and self.atv is None and self._is_on:
+            self._task = asyncio.ensure_future(
+                self._connect_loop(), loop=self.hass.loop
+            )
+        else:
+            _LOGGER.debug(
+                "Not starting connect loop (%s, %s)", self.atv is None, self._is_on
+            )
 
-    tasks = [_setup_atv(hass, config, conf) for conf in config.get(DOMAIN, [])]
-    if tasks:
-        await asyncio.wait(tasks)
+    async def _connect_loop(self):
+        _LOGGER.debug("Starting connect loop")
+
+        # Try to find device and connect as long as the user has said that
+        # we are allowed to connect and we are not already connected.
+        while self._is_on and self.atv is None:
+            try:
+                conf = await self._scan()
+                if conf:
+                    await self._connect(conf)
+            except exceptions.AuthenticationError:
+                self._auth_problem()
+                break
+            except asyncio.CancelledError:
+                pass
+            except Exception:
+                _LOGGER.exception("Failed to connect")
+                self.atv = None
+
+            if self.atv is None:
+                self._connection_attempts += 1
+                backoff = min(
+                    randrange(2 ** self._connection_attempts), BACKOFF_TIME_UPPER_LIMIT
+                )
 
-    hass.services.async_register(
-        DOMAIN, SERVICE_SCAN, async_service_handler, schema=APPLE_TV_SCAN_SCHEMA
-    )
+                _LOGGER.debug("Reconnecting in %d seconds", backoff)
+                await asyncio.sleep(backoff)
 
-    hass.services.async_register(
-        DOMAIN,
-        SERVICE_AUTHENTICATE,
-        async_service_handler,
-        schema=APPLE_TV_AUTHENTICATE_SCHEMA,
-    )
+        _LOGGER.debug("Connect loop ended")
+        self._task = None
 
-    return True
+    def _auth_problem(self):
+        _LOGGER.debug("Authentication error, reconfigure integration")
```

Comment:
Wouldn't warning or error make more sense for this loggin message?

Reply:
No, this is purely for debugging purposes. I have added a bunch of these to simplify for myself in case of problem reports. The user will be informed via a persistent notification. Spamming the log with this isn't necessary.

Reply:
Sounds good.